File: src/lib/system-info.ts
/**
 * Hardware detection — drives the model-tier routing.
 *
 * In Tauri production this is replaced by a Rust-side IPC command that uses
 * the `sysinfo` crate. The browser-dev path here exists so `pnpm dev` still
 * produces sensible defaults on Linux/Windows.
 *
 * See spec §2.11 for the tier table.
 */

import type { SystemSnapshot } from "./types";
import os from "node:os";

const GB = 1024 * 1024 * 1024;

export function detectSystem(): SystemSnapshot {
  const platform = os.platform();
  const totalRamGb = Math.round(os.totalmem() / GB);
  // We don't try to estimate free disk from Node — assume the install
  // wouldn't have completed without disk. Tauri override will be accurate.
  const freeDiskGb = 100;

  const platformKind =
    platform === "darwin"
      ? "macos"
      : platform === "win32"
        ? "windows"
        : platform === "linux"
          ? "linux"
          : "unknown";

  const chip = inferChip(platform);
  const tier = computeTier({ chip, totalRamGb });

  return {
    os: platformKind,
    chip,
    totalRamGb,
    freeDiskGb,
    recommendedTier: tier.tier,
    recommendedModel: tier.model,
  };
}

function inferChip(platform: NodeJS.Platform): SystemSnapshot["chip"] {
  if (platform !== "darwin") {
    // Best-effort. Tauri's Rust side does the real work in production.
    if (platform === "win32") return "windows_x64";
    if (platform === "linux") return "linux_x64";
    return "unknown";
  }
  // On macOS we'd shell out to `sysctl hw.model` here. The Rust core does it
  // properly. For dev, we collapse to a coarse heuristic.
  const arch = os.arch();
  if (arch === "arm64") return "apple_m3_or_better";
  return "intel_mac";
}

function computeTier(input: {
  chip: SystemSnapshot["chip"];
  totalRamGb: number;
}): { tier: SystemSnapshot["recommendedTier"]; model: string } {
  const DEFAULT = process.env.DOCKET_MODEL_DEFAULT ?? "qwen3:32b-instruct-q4_K_M";
  const FALLBACK = process.env.DOCKET_MODEL_FALLBACK ?? "qwen3:8b-instruct-q4_K_M";

  if (input.totalRamGb < 16 || input.chip === "intel_mac") {
    return { tier: "unsupported", model: FALLBACK };
  }
  // Highest tier: 32GB+ on M2 Pro or better, or any Linux/Windows with 32GB+.
  const highTierMacs: SystemSnapshot["chip"][] = ["apple_m2_pro_or_better", "apple_m3_or_better"];
  const isHighTierMac = highTierMacs.includes(input.chip);
  const isHighTierOther = input.chip === "windows_x64" || input.chip === "linux_x64";
  if (input.totalRamGb >= 32 && (isHighTierMac || isHighTierOther)) {
    return { tier: "default", model: DEFAULT };
  }
  return { tier: "fallback", model: FALLBACK };
}

File: src-tauri/src/system_info.rs
//! Hardware detection for model-tier routing.

use serde::Serialize;
use sysinfo::System;

#[derive(Serialize)]
pub struct SystemSnapshot {
    pub os: String,
    pub chip: String,
    pub total_ram_gb: u64,
    pub free_disk_gb: u64,
    pub recommended_tier: String,
    pub recommended_model: String,
}

pub fn detect() -> SystemSnapshot {
    let mut sys = System::new_all();
    sys.refresh_all();
    let total_ram_gb = sys.total_memory() / (1024 * 1024 * 1024);

    let chip = inferred_chip();
    let (tier, model) = compute_tier(&chip, total_ram_gb);
    SystemSnapshot {
        os: std::env::consts::OS.to_string(),
        chip,
        total_ram_gb,
        free_disk_gb: 100, // TODO[scott]: real free-disk from `sysinfo::Disks`.
        recommended_tier: tier,
        recommended_model: model,
    }
}

fn inferred_chip() -> String {
    if cfg!(target_os = "macos") && cfg!(target_arch = "aarch64") {
        "apple_silicon".into()
    } else if cfg!(target_os = "macos") {
        "intel_mac".into()
    } else if cfg!(target_os = "windows") {
        "windows_x64".into()
    } else {
        "linux_x64".into()
    }
}

fn compute_tier(chip: &str, ram_gb: u64) -> (String, String) {
    const DEFAULT: &str = "qwen3:32b-instruct-q4_K_M";
    const FALLBACK: &str = "qwen3:8b-instruct-q4_K_M";
    if ram_gb < 16 || chip == "intel_mac" {
        return ("unsupported".into(), FALLBACK.into());
    }
    if ram_gb >= 32 && chip != "intel_mac" {
        return ("default".into(), DEFAULT.into());
    }
    ("fallback".into(), FALLBACK.into())
}
